src/helpers/redirects.ts

```diff
@@ -65,6 +71,143 @@ export const generateStaticRedirects = ({
   }
 }
 
+/**
+ * Routes that match middleware need to always use the SSR function
+ * This generates a rewrite for every middleware in every locale, both with and without a splat
+ */
+const generateMiddlewareRewrites = ({ basePath, middleware, i18n, buildId }) => {
+  const handlerRewrite = (from: string) => ({
+    from: `${basePath}${from}`,
+    to: HANDLER_FUNCTION_PATH,
+    status: 200,
+  })
+
+  return (
+    middleware
+      .map((route) => {
+        const unlocalized = [handlerRewrite(`${route}`), handlerRewrite(`${route}/*`)]
+        if (i18n?.locales?.length > 0) {
+          const localized = i18n.locales.map((locale) => [
+            handlerRewrite(`/${locale}${route}`),
+            handlerRewrite(`/${locale}${route}/*`),
+            handlerRewrite(`/_next/data/${buildId}/${locale}${route}/*`),
+          ])
+          // With i18n, all data routes are prefixed with the locale, but the HTML also has the unprefixed default
+          return [...unlocalized, ...localized]
+        }
+        return [...unlocalized, handlerRewrite(`/_next/data/${buildId}${route}/*`)]
+      })
+      // Flatten the array of arrays. Can't use flatMap as it might be 2 levels deep
+      .flat(2)
+  )
+}
+
+const generateStaticIsrRewrites = ({
+  staticRouteEntries,
+  basePath,
+  i18n,
+  buildId,
+  middleware,
+}: {
+  staticRouteEntries: Array<[string, SsgRoute]>
+  basePath: string
+  i18n: NextConfig['i18n']
+  buildId: string
+  middleware: Array<string>
+}) => {
```

Comment:
```suggestion
}): { 
  staticRoutePaths: Set<string>, 
  staticIsrRoutesThatMatchMiddleware: Array<string>,
  staticIsrRewrites : NetlifyConfig['redirects']
} | undefined => {
```

Suggestion for the return type! It does make the function definition a lil big